### ISSUE TITLE
quantization

### ISSUE THREAD

**Issue by leonzgtee** [NervanaSystems/distiller#316]
can i use this tool to quantize a self-defined model, for example, a mobilenet v2 with deconv op or upsample operation

**Comment by levzlotnik** [comment]
Hi @leonzgtee ,

The way we quantize models is by taking from them each leaf module we want to quantize, creating a wrapper around them and replacing the leaf module with the respective wrapper.  
For example, if we'd want to quantize a model with `nn.Conv2d`s, we'd create a wrapper class around it like `QuantizedConv2d`, and replace the original `nn.Conv2d` object by the `QuantizedConv2d` wrapper object.  
Fortunately, we have already implemented this for most modules, as you can see in the implementation of `PostTrainLinearQuantizer`. There you can see that the `quantizer` holds a dictionary `replacement_factory`, which takes a type of module as a key and outputs a function (which takes the module itself, its name and several configuration arguments) that will create a wrapper around the module and replace with it in the original model.  
https://github.com/NervanaSystems/distiller/blob/112163eb8eb8f53bc7783250b5d1e4f41109e515/distiller/quantization/range_linear.py#L875-L876
So quantizing `nn.ConvTransposed2d` would usually be done by adding 
```
quantizer.replacement_factory[nn.ConvTransposed2d] = replace_param_layer
```
However the class `RangeLinearQuantParamLayerWrapper` doesn't accept a `nn.ConvTransposed2d` layer yet (will raise `ValueError`). A quick work around is to add in your revision the `nn.ConvTransposed2d` layer into the acceptable `RangeLinearQuantParamLayerWrapper` layers in the code, you should be fine since deconv op is a linear op.

If you want to quantize using a `yaml` configuration, I suggest writing a custom quantizer class in the same way described above.

Let us know if you have more question.
Lev

**Comment by leonzgtee** [comment]
thanks for your reply. but there are still something unclear for me . i conduct some research on the quantization of centerNet, a fast model used for object detection .it's backbone net is mobilenet v2 and contain some extra operations such as postproces,  if i want to quantize this model, can i transfer the whle model into the tool. besides, i have noticed that the tool use cifar and imagenet to evalute the performance of a quantized model, how can i use a coco format data to evaluate my model. looking forward to your reply.

**Comment by levzlotnik** [comment]
The `compress_classifier.py` indeed uses only cifar10/imagenet datasets and only compresses classification models. However the framework is valid for object detection models as well, and the code should be pretty similar to `compress_classifier.py`.  
I am actually working on adding an example of compressing object detection models, should be done soon enough, and you could use the sample for your research :)  
In the meantime, you could write a script for quantizing and evaluating your model using the `PostTrainLinearQuantizer` class.  
The loop of quantizing the model is:  
1. Implementing the model by using a separate module for each operation in `.forward` - this is to allow the quantizer to actually find the operations and be able to quantize them.  
For example:
if the original forward was
```
def forward(self, x):
     x = self.conv1(x)
     x = F.relu(x)
     ...
``` 
You should add `nn.ReLU` in the ctor like `self.relu1 = nn.ReLU()` and in forward call the module function:
```
def forward(self, x):
     x = self.conv1(x)
     x = self.relu1(x)
     ...
``` 
2. Define the quantizer - we have plenty of quantizers implemented, chose the one you need and provide it with the configuration of bit widths for activations, weights and accumulator. Then call `quantizer.prepare_model()` and the model is transformed into its quantized counterpart.
3. Train/Evaluate the model using the regular loop for a non quantized model.  

For more info, check out our [docs](https://nervanasystems.github.io/distiller/design.html#quantization) and [examples](https://github.com/NervanaSystems/distiller/tree/master/examples).

Thanks,
Lev

**Comment by Slicerkao** [comment]
Hi,
I think I get a relevant but somewhat fundamental question. 
In my experiment, I first pretrain my own vgg16 with batch norm on cifar10, which reaches acc 90%.
Then, I use the following code to quantize my model:

```
 # x_dat is the input data of a batch. 
 model.eval()
 quantizer = quantization.PostTrainLinearQuantizer(model, bits_activations=32, bits_parameters = 32)
 quantizer.prepare_model(distiller.get_dummy_input(dataset = None, device = None, input_shape = x_dat.shape)) 
```
out of my expectation, when I evaluate on the test data, the acc drop to 0.1. I expect that 32 bit quantization should leads to the original accuracy. When I try bits to be 8, the accuracy is almost the same as the original one. 

Am I ignoring any setup for this default quantizer ?

**Comment by lehahoang** [comment]
Hi @levzlotnik 

Thanks for your clear description of how quantization works in Distiller. I have been looking for this description on your Documentation page for days! 
I only have one humble request if you could group all quantization-related things in one category in your documentation page. I think it would be convenient for dummies like me to search for key information on your documentation. In my opinion, the Design section on the list should be expanded so that readers can easily look up.
Thanks.

**Comment by levzlotnik** [comment]
Hi @leonzgtee ,

I indeed think you're right, we should clarify the things I've explained here. We've talked about it in our team and we agreed we'll add another FAQ section and probably more clarity to this design.
Thank you for your remark, we'll probably update the docs sometime near in the future.

Thanks,
Lev

**Comment by RizhaoCai** [comment]
> The `compress_classifier.py` indeed uses only cifar10/imagenet datasets and only compresses classification models. However the framework is valid for object detection models as well, and the code should be pretty similar to `compress_classifier.py`.
> I am actually working on adding an example of compressing object detection models, should be done soon enough, and you could use the sample for your research :)
> In the meantime, you could write a script for quantizing and evaluating your model using the `PostTrainLinearQuantizer` class.
> The loop of quantizing the model is:
> 
> 1. Implementing the model by using a separate module for each operation in `.forward` - this is to allow the quantizer to actually find the operations and be able to quantize them.
>    For example:
>    if the original forward was
> 
> ```
> def forward(self, x):
>      x = self.conv1(x)
>      x = F.relu(x)
>      ...
> ```
> 
> You should add `nn.ReLU` in the ctor like `self.relu1 = nn.ReLU()` and in forward call the module function:
> 
> ```
> def forward(self, x):
>      x = self.conv1(x)
>      x = self.relu1(x)
>      ...
> ```
> 
> 1. Define the quantizer - we have plenty of quantizers implemented, chose the one you need and provide it with the configuration of bit widths for activations, weights and accumulator. Then call `quantizer.prepare_model()` and the model is transformed into its quantized counterpart.
> 2. Train/Evaluate the model using the regular loop for a non quantized model.
> 
> For more info, check out our [docs](https://nervanasystems.github.io/distiller/design.html#quantization) and [examples](https://github.com/NervanaSystems/distiller/tree/master/examples).
> 
> Thanks,
> Lev

Hello! Thanks for your efforts of working out examples of detection models.
The distiller is awesome and I am also looking for a proper way to use distiller to prune my detection models. Look forward to your update

**Comment by guyjacob** [comment]
@lehahoang, I added a section on preparing a model for quantization to our docs. You can find it [here](https://nervanasystems.github.io/distiller/prepare_model_quant.html). I hope you and others find it useful. I'm closing this issue for now.

(It's another separate section :) But it's top-level so should be easy to find. It's also linked from the README and FAQ. I think our docs need a general refactoring, but that's something for another day...)

**Comment by nzmora** [comment]
@guyjacob @levzlotnik @leonzgtee 
I added @leonzgtee's question and @levzlotnik's reply to the FAQ.
Cheers
Neta
